registry/quay.go

```diff
@@ -0,0 +1,73 @@
+package registry
+
+import (
+	"log"
+	"os"
+	"strings"
+
+	"github.com/coreos/go-quay/quay"
+	"github.com/coreos/go-quay/quay/tag"
+	"github.com/go-openapi/runtime"
+	"github.com/go-openapi/strfmt"
+)
+
+const (
+	quayTokenEnvVarName = "QUAY_AUTH_TOKEN"
+)
+
+// QuayRegistry implements the Registry interface.  It includes a Client, Auth
+// and helper methods for interacting with the quay.io api
+type QuayRegistry struct {
+	Client *quay.Client
+	Auth   QuayAuth
+}
+
+// NewQuayRegistry returns a pointer to a new QuayRegistry
+func NewQuayRegistry(c *quay.Client, auth QuayAuth) *QuayRegistry {
+	return &QuayRegistry{
+		Client: c,
+		Auth:   auth,
+	}
+}
+
+// QuayAuth is an interface providing AuthenticateRequest
+type QuayAuth interface {
+	AuthenticateRequest(runtime.ClientRequest, strfmt.Registry) error
+}
+
+// NewQuayAuth returns a QuayAuth instance
+func NewQuayAuth() QuayAuth {
+	return &realQuayAuth{}
+}
+
+type realQuayAuth struct{}
+
+func (rqa *realQuayAuth) AuthenticateRequest(cr runtime.ClientRequest, reg strfmt.Registry) error {
+	quayTkn := os.Getenv(quayTokenEnvVarName)
+	if quayTkn == "" {
+		log.Fatalf("'%s' env var required", quayTokenEnvVarName)
+	}
+	return cr.SetHeaderParam("Authorization", quayTkn)
+}
+
+// CheckExistence is the ExistenceChecker for QuayRegistry
+func (q *QuayRegistry) CheckExistence(imageAndTag ImageAndTag) error {
+	quayListTagImagesParams := &tag.ListTagImagesParams{
+		Repository: imageAndTag.Image,
+		Tag:        imageAndTag.Tag,
+	}
+
+	_, err := q.Client.Tag.ListTagImages(quayListTagImagesParams, q.Auth)
+	if err != nil {
+		if strings.Contains(err.Error(), "Not Found") {
+			return ErrTagNotFound{
+				imageAndTag: imageAndTag,
+				registry:    "quay.io",
+			}
+		} else if strings.Contains(err.Error(), "write on closed buffer") {
```

Comment:
Hm, I still don't get why we want to ignore the case when we can't check if a tag exists. If we ignore it and then continue, we're basically lying to the caller and telling them the tag existed when we have no idea.

We could, instead, return a specific (exported) error type that callers can check for and act accordingly (i.e. with logic that they use to deal with cases when the check failed and they don't know if the tag exists).

Thoughts?